src/gt4py/next/iterator/transforms/inline_lifts.py

```diff
@@ -139,125 +164,208 @@ class Flag(enum.IntEnum):
         ):
-            stencil = node.fun.args[0]  # type: ignore[attr-defined] # node already asserted to be of type ir.FunCall
-            eligible_lifted_args = [
-                _is_lift(arg) and self.predicate(arg, is_scan_pass_context) for arg in node.args
-            ]
-
-            if isinstance(stencil, ir.Lambda) and any(eligible_lifted_args):
-                # TODO(tehrengruber): we currently only inlining opcount preserving, but what we
-                #  actually want is to inline whenever the argument is not shifted. This is
-                #  currently beyond the capabilities of the inliner and the shift tracer.
-                new_arg_exprs: dict[ir.Sym, ir.Expr] = {}
-                inlined_args = []
-                for i, (arg, eligible) in enumerate(zip(node.args, eligible_lifted_args)):
-                    if eligible:
-                        assert isinstance(arg, ir.FunCall)
-                        inlined_arg, _ = _transform_and_extract_lift_args(
-                            arg, symtable, new_arg_exprs
+            return None
+
+        stencil = node.fun.args[0]  # type: ignore[attr-defined] # node already asserted to be of type ir.FunCall
+
+        if not isinstance(stencil, ir.Lambda):
+            return None
+
+        eligible_lifted_args = [False] * len(node.args)
+        for i, (param, arg) in enumerate(zip(stencil.params, node.args, strict=True)):
+            if not _is_lift(arg):
+                continue
+            # we don't want to inline lift args derefed at multiple positions as this would
+            # disallow creating a temporary for them if the (outer) lift can not be extracted into
+            # a temporary.
+            if self.inline_single_pos_deref_lift_args_only:
+                if param.annex.recorded_shifts not in [set(), {()}]:
+                    continue
+            eligible_lifted_args[i] = True
+
+        if isinstance(stencil, ir.Lambda) and any(eligible_lifted_args):
+            new_arg_exprs: dict[ir.Sym, ir.Expr] = {}
+            inlined_args: dict[ir.Sym, ir.Expr] = {}
+
+            reserved_symbol_names = list(
+                set(symtable.keys()) | {param.id for param in stencil.params}
+            )
+            for param, arg, eligible in zip(
+                stencil.params, node.args, eligible_lifted_args, strict=True
+            ):
+                if eligible:
+                    transformed_arg, _ = _transform_and_extract_lift_args(
+                        arg,  # type: ignore[arg-type] # ensure by _is_lift on arg
+                        reserved_symbol_names,
+                        new_arg_exprs,
+                        getattr(param.annex, "recorded_shifts", None),
+                    )
+                    inlined_args[param] = transformed_arg
+                else:
+                    # the else-branch is completely sufficient, but this makes the resulting
```

Comment:
```suggestion
                    # the true-branch is completely sufficient, but this makes the resulting
```
or I don't understand what you mean

Reply:
I mean the `false`-branch. Everything would just work with the one-liner there and it would be much nicer to not have the true-branch at all since. I can reformulate to `true`-branch is superfluous instead of `false`-branch is sufficient. As you like.